src/api/route-services/access/index.js

```diff
@@ -0,0 +1,100 @@
+
+const _ = require('lodash');
+
+const config = require('../../../config');
+const AWS = require('../../../utils/requireAWS');
+const getLogger = require('../../../utils/getLogger');
+
+const {
+  createDynamoDbInstance,
+  convertToJsObject,
+  convertToDynamoDbRecord,
+} = require('../../../utils/dynamoDb');
+
+const { isRoleAuthorized } = require('./roles');
+
+const logger = getLogger('[AccessService] - ');
+
+class AccessService {
+  constructor() {
+    this.userAccessTableName = `user-access-${config.clusterEnv}`;
+    this.inviteAccessTableName = `invite-access-${config.clusterEnv}`;
+  }
+
+  // grantRole adds or replaces role for userId and experimentId
+  async grantRole(userId, experimentId, projectId, role) {
+    logger.log(`PUT experiment role ${experimentId} data`);
+
+    const params = {
+      TableName: this.userAccessTableName,
+      Item: {
+        userId,
+        experimentId,
+        projectId,
+        role,
+      },
+    };
+
+    const docClient = new AWS.DynamoDB.DocumentClient({
+      region: config.awsRegion,
+    });
+
+    await docClient.put(params).promise();
+  }
+
+  async canAccessExperiment(userId, experimentId, url, method) {
+    logger.log(`GET experiment permissions ${experimentId} data`);
+    const key = convertToDynamoDbRecord({
+      experimentId,
```

Comment:
what will happen in the `canAccessExperiment` if we pass in nonexistent `experimentId` or nonexistent `userId`?

Reply:
then `const { role } = convertToJsObject(response.Item);` will return `undefined` and `isRoleAuthorized(role, url, method)` will correctly reject the authorization

Reply:
wouldn't it be better if we make it so that if `role` is undefined, we reject the authorization directly in `canAccessExperiment` before we've called `isRoleAuthorized`? We know that having `undefined` for role will never be valid, why continuing futher?

Reply:
yes, it did something like that